native/utils/src/lib.rs

```diff
@@ -34,3 +48,151 @@ pub fn down_cast_any_ref(any: &dyn Any) -> &dyn Any {
         any
     }
 }
+
+/// Preprocesses input arrays to add timezone information from Spark to Arrow array datatype or
+/// to apply timezone offset.
+//
+//  We consider the following cases:
+//
+//  | --------------------- | ------------ | ----------------- | -------------------------------- |
+//  | Conversion            | Input array  | Timezone          | Output array                     |
+//  | --------------------- | ------------ | ----------------- | -------------------------------- |
+//  | Timestamp ->          | Array in UTC | Timezone of input | A timestamp with the timezone    |
+//  |  Utf8 or Date32       |              |                   | offset applied and timezone      |
+//  |                       |              |                   | removed                          |
+//  | --------------------- | ------------ | ----------------- | -------------------------------- |
+//  | Timestamp ->          | Array in UTC | Timezone of input | Same as input array              |
+//  |  Timestamp  w/Timezone|              |                   |                                  |
+//  | --------------------- | ------------ | ----------------- | -------------------------------- |
+//  | Timestamp_ntz ->      | Array in     | Timezone of input | Same as input array              |
+//  |   Utf8 or Date32      | timezone     |                   |                                  |
+//  |                       | session local|                   |                                  |
+//  |                       | timezone     |                   |                                  |
+//  | --------------------- | ------------ | ----------------- | -------------------------------- |
+//  | Timestamp_ntz ->      | Array in     | Timezone of input |  Array in UTC and timezone       |
+//  |  Timestamp w/Timezone | session local|                   |  specified in input              |
+//  |                       | timezone     |                   |                                  |
+//  | --------------------- | ------------ | ----------------- | -------------------------------- |
+//  | Timestamp(_ntz) ->    |                                                                     |
+//  |        Any other type |              Not Supported                                          |
+//  | --------------------- | ------------ | ----------------- | -------------------------------- |
+//
+pub fn array_with_timezone(
```

Comment:
I think these functions are specified to Spark cast. Maybe they should be with cast expression in spark-expr crate?

Reply:
`array_with_timezone` is also called from some other expressions in `temporal.rs`. Once all the expressions are moved to the new crate then maybe we can move some of these utility functions there as well.

Reply:
I suppose some of these functions could just live in the spark-expr crate. I will look at that as a follow on issue.